src/service/resources-impl.js

```diff
@@ -407,13 +410,29 @@ export class Resources {
    * @private
    */
   buildReadyResources_() {
+    // Avoid cases where elements add more elements inside of them
+    // and cause an infinite loop of building - see #3354 for details.
+    if (this.isCurrentlyBuildingPendingResources_) {
```

Comment:
Yeah, this will work. But it's rather cryptic and fragile. We already considered once to reverse traversal from the end of the array. At the very least, let's comment above the for-loop to make it super clear that this cannot be cached.

Also, I have some concerns about optimizers. I see this pattern very common in our binaries and assume that this is the minifier work, but I could be wrong, of course:
```
for (var d = 0, e = a.length; d < e; d++)
```